lib/pos_consensus.go

```diff
@@ -0,0 +1,843 @@
+		return errors.Errorf("Unexpected event type: %v vs %v", event.EventType, expectedEventType)
+	}
+
+	// Validate that the event is properly formed
+	if !isProperlyFormedBlockProposalEvent(event) {
+		return errors.Errorf("Received improperly formed block construction event: %v", event)
+	}
+
+	// Fetch the parent block
+	parentBlockHash := BlockHashFromConsensusInterface(event.QC.GetBlockHash())
+	parentBlock, parentBlockExists := cc.blockchain.blockIndexByHash[*parentBlockHash]
+	if !parentBlockExists {
+		return errors.Errorf("Error fetching parent block: %v", parentBlockHash)
+	}
+
+	// Make sure that the parent has a validated status. This should never fail. If it does, something is
+	// very wrong with the safeBlocks parameter in the FastHotStuffEventLoop.
+	if !parentBlock.IsValidated() {
+		return errors.Errorf("Parent block is not validated: %v", parentBlockHash)
+	}
+
+	// Perform simple block height and view sanity checks on the block construction signal.
+
+	// Cross-validate that the event's tip block height matches the parent's height. If the two don't match
+	// then something is very wrong.
+	if uint64(parentBlock.Height) != event.TipBlockHeight {
+		return errors.Errorf(
+			"Error constructing block at height %d. Expected block height %d",
+			event.TipBlockHeight+1,
+			parentBlock.Height+1,
+		)
+	}
+
+	// Validate that the event's view is not stale. If the view is stale, then it means that the consensus
+	// has advanced to the next view after queuing this block proposal event. This is normal and an expected
+	// race condition in the steady-state.
+	currentView := cc.fastHotStuffEventLoop.GetCurrentView()
+	if currentView > event.View {
+		return errors.Errorf(
+			"Error constructing block at height %d. Stale view %d",
+			event.TipBlockHeight+1,
+			event.View,
+		)
+	}
+
+	// Compute the random seed hash for the previous block's proposer signature
+	parentBlockRandomSeedHash, err := HashRandomSeedSignature(parentBlock.Header.ProposerRandomSeedSignature)
+	if err != nil {
+		return errors.Wrapf(err, "Error computing random seed hash for block at height %d: ", event.TipBlockHeight+1)
+	}
+
+	// Compute the next proposer random seed signature
+	proposerRandomSeedSignature, err := cc.signer.SignRandomSeedHash(parentBlockRandomSeedHash)
+	if err != nil {
+		return errors.Wrapf(err, "Error signing random seed hash for block at height %d: ", event.TipBlockHeight+1)
+	}
+
+	// Construct the unsigned block
+	unsignedBlock, err := cc.produceUnsignedBlockForBlockProposalEvent(event, proposerRandomSeedSignature)
+	if err != nil {
+		return errors.Wrapf(err, "Error producing unsigned block for proposal at height %d", event.TipBlockHeight+1)
+	}
+
+	// Sign the block
+	blockHash, err := unsignedBlock.Header.Hash()
+	if err != nil {
+		return errors.Errorf("Error hashing block: %v", err)
+	}
+	unsignedBlock.Header.ProposerVotePartialSignature, err = cc.signer.SignBlockProposal(unsignedBlock.Header.ProposedInView, blockHash)
```

Comment:
I don't think there's any way for this to be incorrect but just to be safe I would replace all instances of Header.ProposedInview with GetView(), including this instance here.

Reply:
The GetView() is a convenience when accepting incoming PoW blocks so that we can use block height as a pseudo "view". IMO, it's more appropriate to use the raw ProposedInView field here when constructing and validating blocks. This way, there's no way we can misread and conflate the block height as the view.